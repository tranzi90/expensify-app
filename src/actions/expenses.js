import database from "../firebase"

export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
})

export const startAddExpense = (expenseData = {}) => {
    return async (dispatch, getState) => {
        const uid = getState().auth.uid
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData

        const expense = { description, note, amount, createdAt}

        const ref = await database.ref(`users/${uid}/expenses`).push(expense)

        dispatch(addExpense({
                    id: ref.key,
                    ...expense
                }))
    }
}

export const removeExpense = (id) => ({
    type: 'REMOVE_EXPENSE',
    id
})

export const startRemoveExpense = (id) => {
    return async (dispatch, getState) => {
        const uid = getState().auth.uid
        await database.ref(`users/${uid}/expenses/${id}`).remove()

        dispatch(removeExpense(id))
    }
}

export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
})

export const startEditExpense = (id, updates) => {
    return async (dispatch, getState) => {
        const uid = getState().auth.uid
        await database.ref(`users/${uid}/expenses/${id}`).update(updates)

        dispatch(editExpense(id, updates))
    }
}

export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
})

export const startSetExpenses = () => {
    return async (dispatch, getState) => {
        const uid = getState().auth.uid
        const snapshot = await database.ref(`users/${uid}/expenses`).once('value')

        let expenses = []
        snapshot.forEach((child) => {
            expenses.push({
                id: child.key,
                ...child.val()
            })
        })

        dispatch(setExpenses(expenses))
    }
}