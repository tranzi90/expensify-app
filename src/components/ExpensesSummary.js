import React, {Component} from "react"
import selectExpenses from "../selectors/expenses"
import {connect} from "react-redux"
import {Link} from "react-router-dom"
import getExpensesTotal from "../selectors/expenses-total"

export class ExpensesSummary extends Component {
    render() {
        return (
            <div className="page-header">
                <div className='content-container'>
                    <h1 className='page-header__title'><span>{this.props.expenseCount}</span> expenses on total sum <span>{this.props.expensesTotal}</span>:</h1>
                    <div className="page-header__actions">
                        <Link className="button" to='/create'>Add Expense</Link>
                    </div>
                </div>
            </div>
        )
    }
}

const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters)

    return {
        expenseCount: visibleExpenses.length,
        expensesTotal: getExpensesTotal(visibleExpenses)
    }
}

export default connect(mapStateToProps)(ExpensesSummary)