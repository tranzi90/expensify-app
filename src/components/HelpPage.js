import React from "react"

export default () => (
    <div>
        Help Page
    </div>
)