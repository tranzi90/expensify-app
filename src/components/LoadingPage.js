import React from "react"

export default () => (
    <div className='loader'>
        <img src="/images/loader.gif" className='loader__image' alt=''/>
    </div>
)