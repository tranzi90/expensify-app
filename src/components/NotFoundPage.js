import React from "react"

export default () => (
    <div>
        Nema ce
    </div>
)