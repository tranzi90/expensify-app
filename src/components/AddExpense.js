import {connect} from "react-redux"
import ExpenseForm from "./ExpenseForm"
import {addExpense} from "../actions/expenses"
import React, {Component} from 'react'

export class AddExpense extends Component {
    onSubmit = (expense) => {
        this.props.addExpense(expense)
        this.props.history.push('/')
    }
    render() {
        return (
            <div>
                <ExpenseForm
                    onSubmit={this.onSubmit}
                />
            </div>
        )
    }
}

function mapDispatchToProps(dispatch) {
    return {
        addExpense: (expense) => dispatch(addExpense(expense))
    }
}

export default connect(undefined, mapDispatchToProps)(AddExpense)