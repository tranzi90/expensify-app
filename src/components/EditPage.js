import React, {Component} from 'react'
import {connect} from "react-redux"
import ExpenseForm from "./ExpenseForm"
import {startEditExpense, startRemoveExpense} from "../actions/expenses"

export class EditPage extends Component {
    onSubmit = (expense) => {
        this.props.startEditExpense(this.props.expense.id, expense)
        this.props.history.push('/')
    }
    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Edit Expense</h1>
                    </div>
                </div>
                <div className="content-container">
                    <ExpenseForm
                        expense={this.props.expense}
                        onSubmit={this.onSubmit}
                    />
                    <button
                        className='button button--secondary'
                        onClick={ () => {
                            this.props.startRemoveExpense(this.props.expense.id)
                            this.props.history.push('/')
                        }}
                    >Remove expense</button>
                </div>
            </div>
        )
    }
}

function mapStateToProps(state, props) {
    return {
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
    }
}

function mapDispatchToProps(dispatch) {
    return {
        startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
        startRemoveExpense: (id) => dispatch(startRemoveExpense(id))
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(EditPage)
