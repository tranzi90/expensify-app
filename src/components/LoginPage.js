import React from "react"
import {connect} from "react-redux"
import {startLogin} from "../actions/auth"

export function LoginPage({startLogin}) {
    return (
        <div className='box-layout'>
            <div className="box-layout__box">
                <h1 className="box-layout__title">Expensify</h1>
                <p>Control your budget!</p>
                <button onClick={startLogin} className='button'>Login with Google</button>
            </div>
        </div>
    )
}

function mapDispatchToProps(dispatch) {
    return {
        startLogin: () => dispatch(startLogin())
    }
}

export default connect(undefined, mapDispatchToProps)(LoginPage)
