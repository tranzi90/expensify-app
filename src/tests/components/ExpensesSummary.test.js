import { shallow } from 'enzyme'
import { ExpensesSummary } from '../../components/ExpensesSummary'
import React from 'react'
import expenses from '../fixtures/expenses'
import getExpensesTotal from '../../selectors/expenses-total'

it('should view 2 expenses totalling $114000', () => {
    const wrapper = shallow(
        <ExpensesSummary
            expenseCount={[expenses[1], expenses[2]].length}
            expensesTotal={getExpensesTotal([expenses[1], expenses[2]])}
        />
    )
    expect(wrapper).toMatchSnapshot()
})

it('should view 1 expense totalling $195', () => {
    const wrapper = shallow(
        <ExpensesSummary
            expenseCount={[expenses[0]].length}
            expensesTotal={getExpensesTotal([expenses[0]])}
        />
    )
    expect(wrapper).toMatchSnapshot()
})
