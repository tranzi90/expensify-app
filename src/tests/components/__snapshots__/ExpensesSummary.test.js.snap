// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should view 1 expense totalling $195 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      <span>
        1
      </span>
       expenses on total sum 
      <span>
        195
      </span>
      :
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should view 2 expenses totalling $114000 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      <span>
        2
      </span>
       expenses on total sum 
      <span>
        114000
      </span>
      :
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;
